tests/miral/fd_manager.cpp

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright © 2022 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "fd_manager.h"
+
+#include <mir/test/doubles/mock_main_loop.h>
+
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+
+using namespace testing;
+
+using namespace mir::test::doubles;
+
+TEST(FdManager, dropping_fd_handle_after_main_loop_created_unregisters_handler)
+{
+    auto const manager = std::make_shared<miral::FdManager>();
+    auto const main_loop = std::make_shared<NiceMock<MockMainLoop>>();
```

Comment:
I believe `FdManager` will need to remain a `shared_ptr` here because it inherits `std::enable_shared_from_this` so `FdHandleImpl` can keep the `FdManager` alive until all handles are destroyed.

Reply:
OK, although like most use of `enable_shared_from_this` that's ugly and should probably not be combined with an accessible constructor.

I still think a test fixture would unclutter the test code